src/dcqc/parsers.py

```diff
@@ -45,10 +52,83 @@ def create_suites(
         skipped_tests: Optional[Collection[str]] = None,
     ) -> Iterator[SuiteABC]:
         for target in self.create_targets():
-            # This function assumes that there is one file per target,
-            # which is always the case when parsing CSV files (for now)
-            first_file = target.files[0]
-            file_type = first_file.get_file_type()
-            suite_cls = SuiteABC.get_subclass_by_file_type(file_type)
-            suite = suite_cls(target, required_tests, skipped_tests)
-            yield suite
+            yield SuiteABC.from_target(target, required_tests, skipped_tests)
+
+
+class JsonParser:
+    path: Path
+
+    def __init__(self, path: Path):
+        self.path = path
+
+    @classmethod
+    def parse_expected(cls, path: Path, expected_cls: Type[T]) -> T:
+        """Generate expected object from JSON file.
+
+        Args:
+            json_path: JSON file describine the expected object.
+
+        Raises:
+            ValueError: If the JSON file does not
+                describe a expected object.
+
+        Returns:
+            The reconstructed object.
+        """
+        parser = JsonParser(path)
+        object_ = parser.parse_object()
+
+        if not isinstance(object_, expected_cls):
+            message = f"Parsed JSON file ({path!s}) does not describe a {expected_cls}."
+            raise ValueError(message)
+
+        return object_
+
+    def load_json(self) -> Any:
+        with self.path.open("r") as infile:
+            contents = json.load(infile)
+        return contents
+
+    def get_class(self, cls_name: str) -> Type[SerializableMixin]:
```

Comment:
There is a difference. 

First, `Type` and `TypeVar` (used to create `T`) mean different things. You can check out the `typing` docs for more information, but the examples below will demonstrate their usage. 

Second, the `Type[SomeClass]` annotation informs the static type checker to expect `SomeClass` (sub)class objects instead of `SomeClass` (sub)class instances. 

Third, when you generate a type variable using `TypeVar` (_e.g._ `T`, which is a conventional shorthand), you can use this value to inform the type checker that its value is consistent throughout the function signature. I just added a function [here](https://github.com/Sage-Bionetworks-Workflows/py-dcqc/blob/df99b50df1b2620aebbd496d7f67ed67cf8bc56a/src/dcqc/mixins.py#L68-L79) that uses a type variable (relevant snippet copied below). In this example, the function signature is interpreted by the type checker as follows: the return type will be the same as the type of `self`, and this type must be a instance of `SerializableMixin`. 

```
T = TypeVar("T", bound="SerializableMixin")

def copy(self: T) -> T:
    ...
```

The reason that I can't write `def copy(self: SerializableMixin) -> SerializableMixin:` is that this is interpreted as follows: the return type will be an instance of `SerializableMixin`, and the type of `self` will be an instance of `SerializableMixin`, but there is no relationship between the two types. In other words, this doesn't convey to the type checker that the return type will match the input type. 

Finally, in the case of the `JsonParser`, I have the following method signature. Given that `JsonParser` can produce objects from any `SerializableMixin` subclass, the type checker didn't like the having the return type being `SerializableMixin` because that isn't specific enough. For example, if I use a `TestABC`-specific method (_e.g._ `get_status()`) after deserializing a test using `JsonParser`, the type checker thinks that's an error because the return value is `SerializableMixin`, which doesn't have a `get_status()` method. I wrote the `parse_expected()` method to overcome this, which also enables with runtime type checking. The type checker interprets the following method signature as follows: the return type will be an instance of the class defined by `expected_cls`. 

```python
def parse_expected(cls, path: Path, expected_cls: Type[T]) -> T:
    ...
```

Reply:
Thanks for explanation!